src/python/pants/backend/python/subsystems/setup.py

```diff
@@ -219,6 +214,34 @@ class PythonSetup(Subsystem):
         ),
         advanced=True,
     )
+    lockfile_generator = EnumOption(
+        "--lockfile-generator",
+        default=LockfileGenerator.POETRY,
+        help=(
+            "Whether to use Pex or Poetry with the `generate-lockfiles` goal.\n\n"
+            "Poetry does not work with `[python-repos]` for custom indexes/cheeseshops. If you use "
+            "this feature, you should likely use Pex.\n\n"
+            "Several users have also had issues with how Poetry's lockfile generation handles "
+            "environment markers for transitive dependencies; certain dependencies end up with "
+            "nonsensical environment markers which cause the dependency to not be installed, then "
+            "for Pants/Pex to complain the dependency is missing, even though it's in the "
+            "lockfile. There is a workaround: for `[python].resolves`, manually create a "
+            "`python_requirement` target for the problematic transitive dependencies so that they "
+            "are seen as direct requirements, rather than transitive. For tool lockfiles, add the "
+            "problematic transitive dependency to `[tool].extra_requirements`, e.g. "
+            "`[isort].extra_requirements`. Then, regenerate the lockfile(s) with the "
+            "`generate-lockfiles` goal. Alternatively, use Pex for generation.\n\n"
+            "Finally, installing from a Poetry-generated lockfile is slower than installing from a "
+            "Pex lockfile.\n\n"
```

Comment:
The followup PR will change Pants's default locks to use Pex. I will benchmark there. So I could delete this until then, or revert it in the followup if it proves untrue.